lib/streamlit/hashing.py

```diff
@@ -202,121 +202,128 @@ def _to_bytes(self, obj, context):
+                    # (e.g. during development).
+                    return self.to_bytes("%s.%s" % (obj.__module__, obj.__name__))
+
+                h = hashlib.new(self.name)
+                # TODO: This may be too restrictive for libraries in development.
+                if os.path.abspath(obj.__code__.co_filename).startswith(os.getcwd()):
+                    context = _get_context(obj)
+                    if obj.__defaults__:
+                        self._update(h, obj.__defaults__, context)
+                    h.update(self._code_to_bytes(obj.__code__, context))
+                else:
+                    # Don't hash code that is not in the current working directory.
+                    self._update(h, obj.__module__)
+                    self._update(h, obj.__name__)
+                return h.digest()
+            elif inspect.iscode(obj):
+                return self._code_to_bytes(obj, context)
+            elif inspect.ismodule(obj):
+                # TODO: Figure out how to best show this kind of warning to the
+                # user. In the meantime, show nothing. This scenario is too common,
+                # so the current warning is quite annoying...
+                # st.warning(('Streamlit does not support hashing modules. '
+                #             'We did not hash `%s`.') % obj.__name__)
+                # TODO: Hash more than just the name for internal modules.
+                return self.to_bytes(obj.__name__)
+            elif inspect.isclass(obj):
                 # TODO: Figure out how to best show this kind of warning to the
-                # user.
-                st.warning("Streamlit cannot hash an object of type %s." % type(obj))
+                # user. In the meantime, show nothing. This scenario is too common,
+                # (e.g. in every "except" statement) so the current warning is
+                # quite annoying...
+                # st.warning(('Streamlit does not support hashing classes. '
+                #             'We did not hash `%s`.') % obj.__name__)
+                # TODO: Hash more than just the name of classes.
+                return self.to_bytes(obj.__name__)
+            elif isinstance(obj, functools.partial):
+                # The return value of functools.partial is not a plain function:
+                # it's a callable object that remembers the original function plus
+                # the values you pickled into it. So here we need to special-case it.
+                h = hashlib.new(self.name)
+                self._update(h, obj.args)
+                self._update(h, obj.func)
+                self._update(h, obj.keywords)
+                return h.digest()
+            else:
+                try:
+                    # As a last resort, we pickle the object to hash it.
+                    return pickle.dumps(obj, pickle.HIGHEST_PROTOCOL)
+                except:
+                    st.warning("Streamlit cannot hash an object of type %s." % type(obj))
```

Comment:
Also: can we make this warning (and the one below) mention `st.cache` somewhere?

For example:

> Streamlit cannot hash objects of type %s.
> 
> **More information:** to prevent unexpected behavior, Streamlit tries to detect mutations in cached objects so it can alert the user if needed. However, this check is not supported for a small class of objects, including %s.
>
> Please file a bug [here](https://github.com/streamlit/streamlit/issues/new/choose).
>
> To stop this warning from showing in the meantime, try one of the following:
> * **Preferred:** modify your code to avoid using this type of object.
> * Or add the argument `ignore_cache=True` to the `st.cache` decorator.

Reply:
Done